File: iOSDemo/iOSDemo/AdDemo/Native/Resource/Cell/AdCell.m
//
//  AdCell.m
//  iOSDemo
//
//  Created by ltz on 2025/2/7.
//

#import "AdCell.h"

@interface AdCell()

@property (strong, nonatomic) UIView * bgView;

@end

@implementation AdCell

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier {
    if (self == [super initWithStyle:style reuseIdentifier:reuseIdentifier]) {
        self.selectionStyle = UITableViewCellSelectionStyleNone;
        self.backgroundColor = [UIColor whiteColor];
        
        //模拟圆角背景
        self.bgView = [UIView new];
        self.bgView.layer.cornerRadius = 9;
        self.bgView.backgroundColor = UIColor.lightGrayColor;
        self.bgView.layer.masksToBounds = YES;
        [self.contentView addSubview:self.bgView];
    }
    return self;
}
 
- (void)setAdView:(ATNativeADView *)adView {
    
    if (_adView == adView) {
        return;
    }
    
    [_adView removeFromSuperview];
    
    [self.bgView addSubview:adView];
    
    [self.bgView mas_remakeConstraints:^(MASConstraintMaker *make) {
        make.left.mas_equalTo(self.contentView).mas_offset(15);
        make.right.mas_equalTo(self.contentView).mas_offset(-15);
        make.top.mas_equalTo(self.contentView);
        make.bottom.mas_equalTo(self.contentView.mas_bottom);
    }];
      
    [adView mas_remakeConstraints:^(MASConstraintMaker *make) {
        make.top.mas_equalTo(self.bgView);
        make.bottom.mas_equalTo(self.bgView.mas_bottom);
        make.height.mas_equalTo(adView.frame.size.height);
        make.width.mas_equalTo(adView.frame.size.width);
        make.centerX.mas_equalTo(self.bgView.mas_centerX);
    }];
    
    _adView = adView;
}

@end

File: SwiftDemo/SwiftDemo/AdDemo/Native/NativeResources/SelfRender/AdCell.swift
//
//  AdCell.swift
//  SwiftDemo
//
//  Created by ltz on 2025/1/21.
//

import UIKit
import AnyThinkSDK
import SnapKit

class AdCell: UITableViewCell {
    
    private lazy var bgView: UIView = {
        let view = UIView()
        view.layer.cornerRadius = 9
        view.backgroundColor = .lightGray
        view.layer.masksToBounds = true
        return view
    }()
    
    var adView: ATNativeADView? {
        didSet {
            // If the same ad view, do nothing
            if oldValue == adView {
                return
            }
            
            // Remove old ad view
            oldValue?.removeFromSuperview()
            
            if let adView = adView {
                bgView.addSubview(adView)
                
                // Layout bgView
                bgView.snp.remakeConstraints { make in
                    make.left.equalTo(contentView).offset(15)
                    make.right.equalTo(contentView).offset(-15)
                    make.top.equalTo(contentView)
                    make.bottom.equalTo(contentView.snp.bottom)
                }
                
                // Layout adView
                adView.snp.remakeConstraints { make in
                    make.top.equalTo(bgView)
                    make.bottom.equalTo(bgView.snp.bottom)
                    make.height.equalTo(adView.frame.size.height)
                    make.width.equalTo(adView.frame.size.width)
                    make.centerX.equalTo(bgView.snp.centerX)
                }
                
                // Ensure SelfRenderView expands adView (for self-render ads)
                for subview in adView.subviews {
                    if subview is SelfRenderView {
                        subview.snp.remakeConstraints { make in
                            make.edges.equalToSuperview()
                        }
                    }
                }
            }
        }
    }
    
    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        selectionStyle = .none
        backgroundColor = .white
        
        contentView.addSubview(bgView)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}
